rfc: wing sdk specification

Chriscbr opened the pull request:
[Rendered version](https://github.com/winglang/wing/blob/rybickic/sdk-rfc/docs/05-rfcs/2023-01-11-sdk-spec.md)

A draft of our first specification for the Wing SDK resources. While the SDK already has some resources and we can keep adding them on today, we would like to use this spec to make it consider how a larger set of SDK resources will look like, and how they will feel like to be used together, and what kinds of standard interfaces and patterns are needed.

For the first draft, our goal is to focus on API quality and consistency over API comprehensiveness. It's more important that our APIs are reliable, and that our APIs feel like they were all made with the same principles, and that they're all composable in intuitive ways -- than ensuring that every use case is supported.

Each resource has several parts:
- **Props** -  fields that can be passed to initialize and configure the resource. Props can be required or optional - but optional props must have a reasonable default which must be documented.
- **Preflight APIs** - properties and methods that can be accessed on the resource during preflight. Methods are often used to glue resources together, like `Bucket.on_upload(handler)`.
- **Inflight APIs** - properties and methods that can be accessed on the resource during inflight.

Related: #83

Closes #662

*By submitting this pull request, I confirm that my contribution is made under the terms of the 
[Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

Chriscbr commented on rfcs/sdk-spec.md:
```diff
@@ -0,0 +1,834 @@
+* Api (P1) - a REST API
+* Metric (P1) - a metric for monitoring system performance
+* Alarm (P1) - an alarm that triggers when a metric crosses a threshold
+* Service (P1) - a long-running service, similar to AWS ECS, Azure Container Instances, GCP Cloud Run
```
Could you provide an example of how a DNS or SSL resource might be used in a Wing app (or describe a high level use case)? I'd like to better understand they would address functional concerns of cloud development

eladb replied:
I think it's basically just adding support for custom domains with HTTPS both for the `Api` resource and the `Website` resources.

ShaiBer replied:
Sounds right

Chriscbr commented:
If we add `Json` to our language per #1053, then perhaps I can suggest using `Json` in place of `Serializable` in this RFC?

To be honest, I don't think "Serializable" makes a lot of sense as a type on its own since it's so opaque -- I reckon it would make more sense as an interface / trait / protocol of some type. For example:

```js
serializable struct Person {
  name: str;
  age: num;
}

struct ResourceProps {
  name: str;
  data: cloud.Bucket;
}
```

But it's such a long keyword, and anyways, this is a bit out of the scope of the RFC. 😅 

----

More thoughts on event-driven API patterns...

To recap, at a high level, we want some kind of event-based API system or collection of patterns in Wing, because we feel this is a important part of developing applications for the cloud. And we want it to be modular, so that for the most part, any event sources (aka observables, publishers) can be connected to any number of listening resources (aka observers, subscribers) - as long as the resource "matches up" with the event source.

But I noticed what _seems_ like a potential issue with the current design. To see, first let's look at one set of the proposed `on_event` style APIs:

- `Bucket.on_upload()` / `Bucket.on_delete()`
- `Function.on_invoke()`
- `Topic.on_event()`
- `Schedule.on_tick()` / `Schedule.on_cron()`
- `Metric.on_threshold()`
- `Table.on_insert()`

All of the APIs I've listed above can be implemented using a model where an events are **pushed** to the target resource(s), and each event source can be observed by any number of resources at a time. This is an important property, because it means you can easily augment your existing app with new event listeners -- and there is no affect on the original resource, or on other listening resources. The cloud provider is responsible for making this all work reliably.

(This is similar to how most event emitter / observer APIs work in languages like C# and NodeJS, to the best of my understanding).

However, there are a few APIs that don't match up with this model:

- `Api.on_get(route, function)`
- `Queue.on_message(function)`

`Api.on_get` has a similar function signature to the APIs in the first list, but it feels slightly different, since you can typically only attach one function per route on a cloud provider. I feel like "one publisher, any number of subscribers" isn't the right mental model as it is for those in the first list. But at the end of the day it's still alright to think of this as a push-based interaction (the API gateway "pushes" a request to inflight host, which must return a response), so it's might be OK to leave it as is.

`Queue.on_message` is the API I'm concerned about. Letting multiple resources subscribe to a single queue does not typically mean work is partitioned between them equally in the cloud because of differences in how they quickly or not they scale with respect to the size of the queue. We also know that when a message is received by one consumer, it is explicitly held from others. (It's like a potluck with not enough food - if I eat more food, then it's less food you eat 😉). Moreover, a queue is based around a pull-based model, not a push-based model.

Sooo TL;DR in the RFC I've proposed the name `Queue.add_consumer` instead of `Queue.on_message`.

----

Some more thoughts about alternative APIs:

While reading about event APIs in other languages, I noticed an important difference between other languages and Wing. In Wing, we do not need to be able to "delete" or "dispose" of connections between resources. Why? Because all of the connections between event sources and event handlers are made in preflight, so that they can be generated in a static terraform file.

So how is this relevant? One question I've been wondering is whether we need to model event sources with their own kind of class or resource. For example, do we need `bucket.on_upload()` to return some kind of `EventSubscription` object? Well, there's no need to ever call `eventSubscription.cancel()`, so as far as I can tell, we can avoid the overhead for now. 

Another API I considered was creating a standard interface like `IEventSource` or `IObservable` that is implemented by each source of events. It could then be used like so:

```js
// wing
let bucket = new cloud.Bucket();
let bucket_uploads = bucket.uploads(); // IObservable

let fn = new cloud.Function(inflight (event: cloud.BucketEvent) => {
  log.info("file ${event.file} uploaded");
});

fn.subscribe(bucket_uploads);
uploads.on_event(bucket_uploads); // equivalent
```

I'm not sure I see a need for this kind of machinery, so I've opted for the current design where resources just have `on_event_type` methods. But if figure out there are good use cases for this API, then it shouldn't be hard to switch to it. 👍 

A lot of API design is tricky to get right on the first try so I think we just have to build some of these resources, try using them, and see how they feel. 😊

Chriscbr commented on rfcs/sdk-spec.md:
```diff
@@ -0,0 +1,834 @@
+* Bucket (P1) - object storage, similar to AWS S3, Azure Blob Storage, GCP Storage
+* Queue (P1) - a message queue, similar to AWS SQS, Azure Storage Queues, GCP Pub/Sub
+* Function (P1) - a serverless function, similar to AWS Lambda, Azure Functions, GCP Cloud Functions
+* Topic (P1) - a pub/sub topic, similar to AWS SNS, Azure Event Grid, GCP Pub/Sub
```
Feels like we don't have enough data - let's wait on more user feedback before changing the name

eladb commented:
> If we add `Json` to our language per https://github.com/winglang/wing/issues/1053, then perhaps I can suggest using `Json` in place of Serializable in this RFC?

What about binary data? For example, if you store a PNG in a bucket and want to `bucket.get()` it, it should return something like `Buffer` or `Blob`?

I would argue that the non-generic bucket APIs should likely use `Blob` instead of `Json` and that there should be `Json.fromBlob(b)`

eladb commented:
> More thoughts on event-driven API patterns...

Good point about `Queue.on_message(function)`, yes, `queue.add_consumer()` makes sense to me.

And I indeed `Api` events don't technically make sense as on-to-many. For `Api`, this is even more dramatic because it returns a single response that comes from the handler. It's a synchronous call (in the sense that it's not a fire-and-forget event).

So we need to be explicit about the difference between consumer APIs that are "fire-and-forget "notifications and consumer APIs that are "handlers". And ideally we should use two different API patterns for these.

I like the `on_xxx` for the former (fire & forget).

Maybe for `Api` we can just use the express-style API (without `on`):

```js
api.get("/route", handler[, opts...]);
api.post("/boom/bam/:fire", handler[, opts..]);
```

Or we can use `add_get()` and `add_post()`...

eladb commented:
> Some more thoughts about alternative APIs:

Very cool, but seems unneeded at this point

Chriscbr commented on rfcs/sdk-spec.md:
```diff
@@ -0,0 +1,834 @@
+* Counter (P1) - an atomic counter
+* Schedule (P1) - a cron job / scheduled task trigger
+* Website (P1) - a CDN-backed static website
+* Api (P1) - a REST API
```
Added to the list as another resource

Chriscbr commented on rfcs/sdk-resources.md:
```diff
@@ -0,0 +1,569 @@
+When designing APIs for Wing, we try to follow these tenets (unless you know better ones):
+
+- **Focused on functional behavior**: our APIs are designed around the functional aspects that developers care about for building and testing their applications.
+Implementations of resources in the SDK are assumed to be scalable, highly-available, and fault tolerant by default, so that developers do not need to customize security policies or scaling configuration within their application code.
```
Reworded from "assumed to be" to "guaranteed to be" to make it clear that scalability/reliability properties is a design goal

Chriscbr commented on rfcs/sdk-resources.md:
```diff
@@ -0,0 +1,569 @@
+In these cases, the API returns an `Iterator` object with a `next()` method that returns a `Promise` for the next page of results and a `has_next()` method that returns a `Promise` for a boolean indicating whether there are more results to fetch.
+The `Iterator` object also implements the [async iterator protocol in JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols), so it can be used within `for...of` loops.
+
+## Planned resources
```
I've added a note in the RFC that we can implement these as third party resources / wing libraries

Chriscbr commented on rfcs/sdk-resources.md:
```diff
@@ -0,0 +1,505 @@
+
+## Planned resources
+
+Resources planned for MVP:
```
Added an `OnDeploy` resource to the list (P2)

Chriscbr commented on rfcs/sdk-resources.md:
```diff
@@ -0,0 +1,505 @@
+Any non-determinism should be minimized and scoped to sources provided by the user (e.g. by letting the user specify input files or environment variables).
+Non-determistic information should also ideally be managed by the provisioning engine's (for example, random IDs can be generated and managed with Terraform state).
+
+## Concepts used throughout the RFC
```
> I think we also need a programming model that talks about stateful and stateless resources. For example, it should be possible to centrally (and locally) control how stateful resources are replaced across deployments (lifecycle).

This is a big topic to unpack! I've tried to write down some initial thoughts on it [here](https://github.com/winglang/wing/discussions/1054), but I have a feeling you probably have a lot of other/different ideas in mind so I'd be eager to hear them and try to incorporate them in the RFC.

Could you explain what you mean by "control how stateful resources are replaced"? Are you referring to CloudFormation's [UpdateReplacePolicy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatereplacepolicy.html)?  It looks a neat feature -- we could try emulating that in Terraform, but I wonder if it might fall out of our purview as it seems fairly deployment engine related. Could you create an issue about this?

Chriscbr commented on rfcs/sdk-resources.md:
```diff
@@ -0,0 +1,410 @@
+     * The amount of memory to allocate to the function, in MB.
+     * @default 256MB
+     */
+    memory?: num;
```
Left in memory for now, to be pragmatic. I'm not sure if "a prop must be simulatable in order to be in the API" is a good heuristic or not

Chriscbr commented on rfcs/sdk-resources.md:
```diff
@@ -0,0 +1,410 @@
+This includes all primitive types, as well as `List`s and `Map`s of `Serializable`s.
+The Wing language spec does not yet support `Serializable` as a type, but until it is supported we might choose to use the `any` or `str` types instead.
+
+### IEventObserver
```
Added a section about event-based programming and how it works and how users implement it. It does some handwaving about an `EventHandler<T>` interface that doesn't exist yet, but I think that might be OK for now, happy to change it or remove those parts if desired.

Chriscbr commented on rfcs/sdk-resources.md:
```diff
@@ -0,0 +1,411 @@
+## Queue
+
+> Note: AWS SQS imposes a (configurable) visibility timeout on all messages received from a queue.
+> To abstract this away from the user, the implementation in Wing will automatically extend the visibility timeout of a message whenever a worker is processing it for an extended period of time, up to the 12 hour maximum.
```
I've unfortunately punted the research of Queue for now, so the API I've written is a placeholder for now (I left a comment in the document).

I think a Queue is a cool resource, but I am not familiar enough with the queueing theory and how different messaging services operates on different clouds to feel confident with this design, so I think it's best for it to be researched and spec'd out independent of the rest of this RFC - including the questions of whether they should be FIFO, how visibility timeouts are handled, etc.

Chriscbr commented on rfcs/sdk-resources.md:
````diff
@@ -0,0 +1,411 @@
+
+```ts
+interface ApiProps {
+    /**
````
Added a first pass at a CORS API :)

Chriscbr commented on rfcs/sdk-resources.md:
```diff
@@ -0,0 +1,505 @@
+In these cases, the API returns an `Iterator` object with a `next()` method that returns a `Promise` for the next page of results and a `has_next()` method that returns a `Promise` for a boolean indicating whether there are more results to fetch.
+The `Iterator` object also implements the [async iterator protocol in JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols), so it can be used within `for...of` loops.
+
+## Planned resources
```
On further thought, I think a survey might be too low-level for the kind of information we want. Let's look to continue getting community feedback in other ways

eladb commented:
Relevant: https://santhoshkumarravi.medium.com/designing-event-driven-architecture-1681b9ad6e65

staycoolcall911 reviewed (commented):
Great work @Chriscbr!
Found a few typos while reviewing this for tomorrow.

Chriscbr commented on rfcs/sdk-spec.md:
```diff
@@ -0,0 +1,1043 @@
+   * Pops (deletes and returns) a message from the queue.
+   */
+  inflight pop(): Serializable;
+}
```
Yes! good catch 😄

Chriscbr commented on rfcs/sdk-spec.md:
```diff
@@ -0,0 +1,1043 @@
+  /**
+   * Invoke the function.
+   */
+  inflight invoke(payload: Serializable): Serializable;
```
In my current understanding, it can return whatever the FaaS returns. For example, a function could do a quick calculation and return the result to whoever called `invoke`.

It's good you ask this because I realized, this only makes sense in cases where you are invoking the function **synchronously**. (e.g. someone calls `invoke` and awaits the function's response). In other cases, you might invoke a function and it might do some work taking several minutes -- so you just want to fire and forget. (In this case, a successful response just means that AWS Lambda received your invocation request -- but it doesn't contain the result).

I'll propose this in the spec as `invoke_async` -- let me know if this mental model makes sense, and if if you have any better name ideas 🙂

staycoolcall911 replied:
Makes sense, thank you! Maybe you could add a `@returns` annotation to the function signature to explain this behavior to the API users?

Chriscbr replied:
Updated - it should be a bit clearer now.

MarkMcCulloh commented:
Just a heads up, RFCs have been moved from `rfcs/` to `docs/05-rfcs/` (https://github.com/winglang/wing/pull/1133)

hasanaburayyan commented on rfcs/sdk-spec.md:
```diff
@@ -0,0 +1,1043 @@
+   * Returns a presigned url to the given file. This URL can be used to access
+   * the file by anyone until it expires (defaults to 24 hours).
+   */
+  inflight presigned_url(key: str, duration?: duration): str;
```
does calling this inflight `presigned_url` go against cloud/target agnostic? 

To me the naming makes sense but that might just be my AWS bias showing 😅 the same concept in GCP is just called `signed url` and in Azure (who just has to be different) calls this `shared access signature`

Chriscbr replied:
Good callout - it didn't occur to me that presigned url could have been an AWS-ism. On Google, I'm seeing more search results for "signed url" than "presigned url" or "pre-signed url", so I'll change the name 😛

hasanaburayyan commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1043 @@
+  /**
+   * Pops (deletes and returns) a message from the queue.
+   */
+  inflight pop(): Serializable;
```
would it be valuable to add an approximate message count? It looks like I can find a couple times on StackOverflow that devs have been looking for a good way to do this. 
[GCP](https://stackoverflow.com/questions/67329653/how-to-get-task-count-in-gcp-cloud-task)
[AWS](https://stackoverflow.com/questions/25351145/get-number-of-messages-in-an-amazon-sqs-queue)
[AZURE](https://stackoverflow.com/questions/11129761/unable-to-get-queue-length-message-count-from-azure)
```js
/**
* Retrieves approximate message count from queue
*/
inflight messageCount(): number
```

Chriscbr replied:
Yes, totally :) what do you think of `approx_size`, or `size_approx` as an API name?

hasanaburayyan replied:
I think approx_size would be my vote

Chriscbr commented on rfcs/sdk-spec.md:
```diff
@@ -0,0 +1,1043 @@
+struct FunctionProps {
+  /**
+   * The maximum amount of time the function can run, in seconds.
+   * @default 1min
```
Removed the "in seconds" part and fixed the default from `1min` to `1m` instead as that's the current shorthand in the language

yoav-steinberg commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
````diff
@@ -0,0 +1,1070 @@
+
+Example:
+
+```ts
````
```suggestion
```ts
// wing
```

yoav-steinberg commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+struct ApiOnPatchProps { /* elided */ }
+struct ApiOnRequestProps { /* elided */ }
+
+struct ApiCorsProps {
```
Not sure about this name, what's "cors"?

staycoolcall911 replied:
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

Chriscbr replied:
Yeah - I think anyone that needs to use this option will be familiar with the acronym. 

BTW we should see if it's possible to make this work @yoav-steinberg:

```js
// wing

// OK
new Api(
  cors: ApiCorsProps {
    origins: ["https://example.com"],
  }
}

// better
new Api(
  cors: {
    origins: ["https://example.com"],
  }
}
```

yoav-steinberg commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+  headers: Map<str, str>;
+}
+
+struct ApiResponse {
```
Do we need some api here for reading the response iteratively (in case of huge response)?

Chriscbr replied:
Do you mean if the user wants to create an iterable API? That's a pretty good question... I think the solution here would be for the user to implement pagination in their API. Definitely something we can try to make easier for users, I'll add it as a future extension idea

yoav-steinberg commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
````diff
@@ -0,0 +1,1070 @@
+}
+```
+
+## Table
````
Maybe a word on how this is implemented on different clouds (DynamoDB and ??).

Chriscbr replied:
Added

staycoolcall911 commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+
+## Table
+
+The table resource represents a database table that can be used to store and query data.
```
Is this meant to be a relational DB table?

Chriscbr replied:
I think this API will make sense as a NoSQL database - updated the description to clarify. (The API does not expose any methods for doing SQL queries etc.) I've also updated the list of proposed resources with a SqlDatabase based on the previous team discussion. 👍

staycoolcall911 commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
````diff
@@ -0,0 +1,1070 @@
+}
+```
+
+## Planned resources
````
That's what he said :)
https://github.com/winglang/wing/pull/858#discussion_r1046732665

Chriscbr replied:
Added some notes in the concepts area, and in each resource I labeled whether it was stateful or not. One I wasn't sure about was `Service` - for now I guessed it shouldn't be stateful

eladb commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+
+resource Bucket {
+  init(props: BucketProps = {});
+
```
we need `add_object()` preflight

Chriscbr replied:
Done

Chriscbr commented:
We've reached a final comments period! This is still just the first version of the SDK spec, but next Thursday (Jan 19) we will merge this, and further changes will be made through additional issues and pull requests. We'll also take this opportunity to create tracking issues and tasks based on the RFC.

(@yoav-steinberg @staycoolcall911 I will address your feedback shortly 🙂)

flyingImer commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+
+## Planned resources
+
+* Bucket (P1) - object storage, similar to AWS S3, Azure Blob Storage, GCP Storage
```
+1 on "ObjectStore"

"Bucket" sounds too specific to some cloud products, but not all
"Storage" is a bit vague to me, especially we have "Table" and "Key-value store" in mind

Chriscbr replied:
This is useful feedback. Let's commit to `Bucket` for now -- the reasoning being that it's used across multiple clouds (AWS, GCP, and Cloudflare use this term) and I think it's intuitive for many of our initial adopters. That said, we should keep this flexible / stay open to changing this in the future.

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+  }
+
+  // or: on_threshold(handler: EventHandler<ThresholdReachedEvent>)
+  on_threshold(handler: inflight (event: ThresholdReachedEvent) => void): void {
```
I think the second void can be inferred by the compiler (based on type inference on the body of the function) but not necessarily the first one.

BTW, maybe it should be `inflight (event: ThresholdReachedEvent): void` (no arrow) since it's just a type annotation... but I suppose we could define it to be either way in our grammar.

Chriscbr replied:
I removed the second `void` but left it in most other places for emphasis- let me know if you think it should be otherwise.

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+  threshold: num;
+}
+
+resource MyCounter extends cloud.Counter {
```
Fixed

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+## Planned resources
+
+* Bucket (P1) - object storage, similar to AWS S3, Azure Blob Storage, GCP Storage
+* Queue (P1) - a message queue, similar to AWS SQS, Azure Storage Queues, GCP Pub/Sub
```
Yes, of course. Let me remove the mention of GCP Pub/Sub from the first to avoid confusion (it just happens that I think they might both be implemented using GCP Pub/Sub).

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+
+* DynamoDBTable
+* Redis
+* MongoDB
```
Sure, there could be some common API that will be extracted. I'm not familiar enough with them to do it justice here but we can track it / add it as a feature request if you'd like 🙂

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+  /**
+   * Run an inflight whenever an event is published to the topic.
+   */
+  on_event(fn: inflight (event: Serializable) => void, opts: TopicOnPublishProps?): void;
```
Let's add it as an alias (I see value to both names -- and offering multiple ways to do something is okay in our design tenets)

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+
+struct QueueAddConsumerProps { /* elided */ }
+
+resource Queue {
```
Added to the list of "open questions"

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+  description?: str;
+}
+
+struct MetricOnThresholdProps {
```
cool article!

The metric API is mostly modeled based on my understanding of metrics in AWS  - my mental model there is you just write/log some values with timestamps, and AWS will graph them in your console, let you set up basic alarms based on them, yada yada.

Open to feedback on other use cases we should think about. e.g. I know some people might want to set up triggers / alarms based on mathematical functions on metrics... so I'll add that as an extension idea. (Not sure if we want to model that for MVP).

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
````diff
@@ -0,0 +1,1070 @@
+}
+```
+
+Future extensions: support endpoint authorization? cors?
````
Added an example for Api.

I omitted examples to make it easier to iterate on the spec, but we can definitely add more wherever they're needed (or on the tracking issues directly)!

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
````diff
@@ -0,0 +1,1070 @@
+```ts
+// wing
+
+interface EventHandler<T> {
````
Of course

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
````diff
@@ -0,0 +1,1070 @@
+```
+
+<!--
+We could also decide the should be named `handle_event` or `invoke` or `run` or `call` or something else.
````
Ok - updated the description to avoid this general `EventHandler` interface and allude to some kind of structural typing. I didn't add too much detail since I felt it's more of a compiler/language detail

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+// wing
+
+interface EventHandler<T> {
+  inflight handle(event: T);
```
Updated

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+   * Returns a signed url to the given file. This URL can be used by anyone to
+   * access the file until theh link expires (defaults to 24 hours).
+   */
+  inflight signed_url(key: str, duration?: duration): str;
```
See the discussion & my comment here: https://github.com/winglang/wing/pull/858#discussion_r1067551790

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+  /**
+   * Return the approximate message count of the queue.
+   */
+  inflight approx_size: num;
```
Thanks 👍

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+  /**
+   * Invoke the function.
+   */
+  inflight invoke(payload: Serializable): Serializable;
```
Talked a little bit about this with @ekeren. I think long term there is some value to being able to invoke a function directly with any "serializable" type, including maps and structs that don't refer to resources, without having to manually convert things to JSON. But for MVP let's use `Json` as you both suggested. 👍

Chriscbr replied:
I removed references to `Serializable` in the RFC and replaced them all with `Json` and `Blob`, hopefully should make things clearer. :)

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+fast and easy for you to create secure, production-ready container images from
+source code and without a Dockerfile.
+
+<!--
```
Bad habit, yeah let's just include it in the doc.

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
````diff
@@ -0,0 +1,1070 @@
+}
+```
+
+Future extensions: domain and certificate props? support for edge functions?
````
Let's leave this for future research. I think it's OK if this is not supported for beta

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+}
+
+resource Function {
+  init(handler: inflight (input: any): any, props: FunctionProps = {});
```
Updated

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+   * The service's number of replicas.
+   * @default 1
+   */
+  replicas: number?;
```
Added this as a TODO in the Service resource description

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+  /**
+   * Invoke the function, but do not wait for the result.
+   */
+  inflight invoke_async(payload: Serializable): Serializable;
```
Fixed

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+- **Built with jsii**: The Wing SDK is designed first and foremost for the Wing language, but it is compiled with jsii to allow the resources to be created as (preflight) CDK constructs in all jsii-supported programming languages.
+jsii poses restrictions on language features that cannot be idiomatically represented in target languages, and encourages good practices for object-oriented design.
+Features that are specific to Wing (such as inflight functions) may not be supported in other jsii languages.
+
```
I think some of these ideas are already captured, but maybe just not emphasized well enough - I've updated some of the titles to make these points clearer.

I sense the tenets might be something you want to get the wording or emphasis just right to your taste - so please don't hesitate to just commit directly to my branch or submit a PR or suggest exact sentences to use. 😊

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+* Bucket (P1) - object storage, similar to AWS S3, Azure Blob Storage, GCP Storage
+* Queue (P1) - a message queue, similar to AWS SQS, Azure Storage Queues, GCP Pub/Sub
+* Function (P1) - a serverless function, similar to AWS Lambda, Azure Functions, GCP Cloud Functions
+* Topic (P1) - a pub/sub topic, similar to AWS SNS, Azure Event Grid, GCP Pub/Sub
```
@MarkMcCulloh raised a good point that the phrase "Topic" is used in many other contexts, from C#'s Pub/Sub model all the way to Rabbit MQ. Let's wait for more user input before deciding how this should be (re-)named.

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1070 @@
+The queue resource represents a message buffer that can be used to decouple workloads between a set of producers and a set of consumers.
+Using a queue, you can send, store, and receive messages between software components at any volume, without losing messages or requiring other services to be available.
+
+Any number of producers can push messages to the queue, and any number of consumers can pop messages from the queue.
+When a consumer function receives a batch of messages, it is responsible for processing each message in the batch and acknowledging that the message has been processed by deleting it from the queue.
+If a consumer function does not acknowledge a message, the message will be re-delivered to another consumer function.
+
+A queue is not FIFO (first-in-first-out), so messages may be delivered out of order.
```
Discussed a bit in our last community + team meeting. I share your concern as I'm still not in love with this name. I feel like "TaskQueue" could be slight improvement, in addition to one of your suggestions. While I don't think we have a consensus on keeping the name "Queue", we also don't have much of a consensus for any other name, so let's keep this as a placeholder for now. 😅

Chriscbr commented:
I've addressed most of the comments so far! Please don't hesitate to click `unresolve` on a conversation and add a comment if you want to follow up on something - I know resolving a conversation without a response/acknowledgement is a little bit hostile but it was the easiest way to keep track of all of the feedback 😅

flyingImer commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -195,6 +206,7 @@ struct ThresholdReachedEvent {
 * Stream (P2) - a stream of events, similar to AWS Kinesis, Azure Event Hubs, GCP Pub/Sub and Dataflow
 * OnDeploy (P2) - a variation of Function that runs every time the app is deployed
 * GraphQLApi (P2) - a GraphQL API, similar to AWS AppSync
+* SqlDatabase (P2) - a relational database that lets you execute arbitrary SQL queries, similar to AWS RDS, Azure SQL Database, GCP Cloud SQL
```
wondering if `Table` above should be `SqlTable` instead to follow this naming scheme? or the other way around, i.e., making it simply `Database`

flyingImer commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -980,18 +1154,24 @@ struct ServiceResponse {
 ## Table
 
-The table resource represents a database table that can be used to store and query data.
+The table resource represents a NoSQL database table that can be used to store and query data.
```
like it! I was confused, but you might want to update `Table (P2) - a relational database table` in the `Planned Resources` section

flyingImer commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -195,6 +206,7 @@ struct ThresholdReachedEvent {
 * Stream (P2) - a stream of events, similar to AWS Kinesis, Azure Event Hubs, GCP Pub/Sub and Dataflow
 * OnDeploy (P2) - a variation of Function that runs every time the app is deployed
 * GraphQLApi (P2) - a GraphQL API, similar to AWS AppSync
+* SqlDatabase (P2) - a relational database that lets you execute arbitrary SQL queries, similar to AWS RDS, Azure SQL Database, GCP Cloud SQL
```
~~wondering if `Table` above should be `SqlTable` instead, to follow this naming scheme? or the other way around, i.e., making it simply `Database`~~

EDIT: after reviewing the `Table` section below, I found the `Table` description itself is incorrect. Should be `Table (P2) - a NoSQL database table` I believe

Chriscbr replied:
You're right - I'll fix it

flyingImer reviewed (changes requested):
minor updates needed

Chriscbr commented on docs/05-rfcs/2023-01-11-sdk-spec.md:
```diff
@@ -0,0 +1,1250 @@
+* Metric (P1) - a metric for monitoring system performance
+* Alarm (P1) - an alarm that triggers when a metric crosses a threshold
+* Service (P1) - a long-running service, similar to AWS ECS, Azure Container Instances, GCP Cloud Run
+* Table (P2) - a relational database table
```
```suggestion
* Table (P2) - a NoSQL database table
```

staycoolcall911 commented:
Congrats @Chriscbr !!!

eladb commented:
Major milestone for Wing!
